pkg/storegateway/bucket.go

```diff
@@ -1027,14 +950,137 @@ func (s *BucketStore) Series(req *storepb.SeriesRequest, srv storepb.Store_Serie
 		return
 	}
 
-	if err = srv.Send(storepb.NewStatsResponse(stats.postingsFetchedSizeSum + stats.seriesFetchedSizeSum)); err != nil {
+	unsafeStats := stats.export()
+	if err = srv.Send(storepb.NewStatsResponse(unsafeStats.postingsFetchedSizeSum + unsafeStats.seriesFetchedSizeSum)); err != nil {
 		err = status.Error(codes.Unknown, errors.Wrap(err, "sends series response stats").Error())
 		return
 	}
 
 	return err
 }
 
+// synchronousSeriesSet returns seriesSet that contains the requested series. It returns a cleanup func. The cleanup func
+// should be invoked always when non-nil; even when the returned error is non-nil. synchronousSeriesSet must be called
+// while holding s.mtx.RLock(). synchronousSeriesSet will call s.mtx.RUnlock regardless of its return values
+func (s *BucketStore) synchronousSeriesSet(
+	ctx context.Context,
+	req *storepb.SeriesRequest,
+	stats *safeQueryStats,
+	blocks []*bucketBlock,
+	indexReaders map[ulid.ULID]*bucketIndexReader,
+	chunkReaders map[ulid.ULID]*bucketChunkReader,
+	resHints *hintspb.SeriesResponseHints,
+	shardSelector *sharding.ShardSelector,
+	matchers []*labels.Matcher,
+	chunksLimiter ChunksLimiter,
+	seriesLimiter SeriesLimiter,
+) ([]storepb.SeriesSet, func(), error) {
+	var (
+		res      []storepb.SeriesSet
+		cleanups []func()
+		mtx      sync.Mutex
```

Comment:
[nit] To clarify that the mutex protects.

```suggestion
		resMx      sync.Mutex
		res      []storepb.SeriesSet
		cleanups []func()
```